invalid memory address or nil pointer dereference

Hi, I am getting below error when running against http sites which are working well with Kerberos via Chrome or Safari. My test client is same as httpClient under examples.
any help appreciated.

panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x50 pc=0x71bbc6]

goroutine 5 [running]:
testing.tRunner.func1(0xc0000d4100)
        C:/Go/src/testing/testing.go:792 +0x38e
panic(0x78f6c0, 0xac64d0)
        C:/Go/src/runtime/panic.go:513 +0x1c7
gopkg.in/jcmturner/gokrb5.v7/client.(*Client).Key(0xc00005d4d0, 0x8671c0, 0xaef230, 0x0, 0x0, 0xc00024a240, 0x0, 0x1, 0xc000456370, 0x1)
        C:/projects/krb/src/gopkg.in/jcmturner/gokrb5.v7/client/client.go:120 +0x86
gopkg.in/jcmturner/gokrb5.v7/client.setPAData(0xc00005d4d0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        C:/projects/krb/src/gopkg.in/jcmturner/gokrb5.v7/client/ASExchange.go:94 +0xe2
gopkg.in/jcmturner/gokrb5.v7/client.(*Client).Login(0xc00005d4d0, 0xc00024a1a0, 0xc00024a1c0)
        C:/projects/krb/src/gopkg.in/jcmturner/gokrb5.v7/client/client.go:183 +0x177
gopkg.in/jcmturner/gokrb5.v7/spnego.(*SPNEGO).AcquireCred(0xc00024a1a0, 0x801cef, 0x2b)
        C:/projects/krb/src/gopkg.in/jcmturner/gokrb5.v7/spnego/spnego.go:47 +0x36
gopkg.in/jcmturner/gokrb5.v7/spnego.SetSPNEGOHeader(0xc00005d4d0, 0xc000112000, 0x801cef, 0x2b, 0x0, 0x7f4010)
        C:projects/krb/src/gopkg.in/jcmturner/gokrb5.v7/spnego/http.go:153 +0x13a
gopkg.in/jcmturner/gokrb5.v7/spnego.(*Client).Do(0xc00045a9c0, 0xc000112000, 0x0, 0x0, 0x1)
        C:/projects/krb/src/gopkg.in/jcmturner/gokrb5.v7/spnego/http.go:87 +0x2a7
gopkg.in/jcmturner/gokrb5.v7/spnego.(*Client).Do(0xc00045a9c0, 0xc0000d4300, 0x801cef, 0x2b, 0xc00045a9c0)
        C:/projects/krb/src/gopkg.in/jcmturner/gokrb5.v7/spnego/http.go:81 +0x196
krb.login(0x7f1e8a, 0x5, 0x12, 0xc0000dff58, 0x1, 0x1)
        C:/projects/krb/src/krb/login.go:70 +0x4d9
krb.TestLogin(0xc0000d4100)
        C:/projects/krb/src/krb/login_test.go:27 +0x147
testing.tRunner(0xc0000d4100, 0x80d4f0)
        C:/Go/src/testing/testing.go:827 +0xc6
created by testing.(*T).Run
        C:/Go/src/testing/testing.go:878 +0x35a

--- a/client/ASExchange.go
+++ b/client/ASExchange.go
@@ -17,6 +17,12 @@ func (cl *Client) ASExchange(realm string, ASReq messages.ASReq, referral int) (
 		return messages.ASRep{}, krberror.Errorf(err, krberror.ConfigError, "AS Exchange cannot be performed")
 	}
 
+	// Set PAData if required
+	err := setPAData(cl, nil, &ASReq)
+	if err != nil {
+		return messages.ASRep{}, krberror.Errorf(err, krberror.KRBMsgError, "AS Exchange Error: issue with setting PAData on AS_REQ")
+	}
+
 	b, err := ASReq.Marshal()
 	if err != nil {
 		return messages.ASRep{}, krberror.Errorf(err, krberror.EncodingError, "AS Exchange Error: failed marshaling AS_REQ")
@@ -30,7 +36,7 @@ func (cl *Client) ASExchange(realm string, ASReq messages.ASReq, referral int) (
 			case errorcode.KDC_ERR_PREAUTH_REQUIRED, errorcode.KDC_ERR_PREAUTH_FAILED:
 				// From now on assume this client will need to do this pre-auth and set the PAData
 				cl.settings.assumePreAuthentication = true
-				err = setPAData(cl, e, &ASReq)
+				err = setPAData(cl, &e, &ASReq)
 				if err != nil {
 					return messages.ASRep{}, krberror.Errorf(err, krberror.KRBMsgError, "AS Exchange Error: failed setting AS_REQ PAData for pre-authentication required")
 				}
@@ -70,7 +76,7 @@ func (cl *Client) ASExchange(realm string, ASReq messages.ASReq, referral int) (
 }
 
 // setPAData adds pre-authentication data to the AS_REQ.
-func setPAData(cl *Client, krberr messages.KRBError, ASReq *messages.ASReq) error {
+func setPAData(cl *Client, krberr *messages.KRBError, ASReq *messages.ASReq) error {
 	if !cl.settings.DisablePAFXFAST() {
 		pa := types.PAData{PADataType: patype.PA_REQ_ENC_PA_REP}
 		ASReq.PAData = append(ASReq.PAData, pa)
@@ -80,7 +86,7 @@ func setPAData(cl *Client, krberr messages.KRBError, ASReq *messages.ASReq) erro
 		var et etype.EType
 		var err error
 		var key types.EncryptionKey
-		if krberr.ErrorCode == 0 {
+		if krberr == nil {
 			// This is not in response to an error from the KDC. It is preemptive or renewal
 			// There is no KRB Error that tells us the etype to use
 			etn := cl.settings.preAuthEType // Use the etype that may have previously been negotiated
@@ -102,7 +108,7 @@ func setPAData(cl *Client, krberr messages.KRBError, ASReq *messages.ASReq) erro
 				return krberror.Errorf(err, krberror.EncryptingError, "error getting etype for pre-auth encryption")
 			}
 			cl.settings.preAuthEType = et.GetETypeID() // Set the etype that has been defined for potential future use
-			key, err = cl.Key(et, &krberr)
+			key, err = cl.Key(et, krberr)
 			if err != nil {
 				return krberror.Errorf(err, krberror.EncryptingError, "error getting key from credentials")
 			}
@@ -138,7 +144,7 @@ func setPAData(cl *Client, krberr messages.KRBError, ASReq *messages.ASReq) erro
 }
 
 // preAuthEType establishes what encryption type to use for pre-authentication from the KRBError returned from the KDC.
-func preAuthEType(krberr messages.KRBError) (etype etype.EType, err error) {
+func preAuthEType(krberr *messages.KRBError) (etype etype.EType, err error) {
 	//The preferred ordering of the "hint" pre-authentication data that
 	//affect client key selection is: ETYPE-INFO2, followed by ETYPE-INFO,
 	//followed by PW-SALT.

--- a/client/ASExchange.go
+++ b/client/ASExchange.go
@@ -81,13 +81,13 @@ func setPAData(cl *Client, krberr messages.KRBError, ASReq *messages.ASReq) erro
 		var err error
 		var key types.EncryptionKey
 		if krberr.ErrorCode == 0 {
+			// This is not in response to an error from the KDC. It is preemptive or renewal
 			// There is no KRB Error that tells us the etype to use
 			etn := cl.settings.preAuthEType // Use the etype that may have previously been negotiated
 			if etn == 0 {
 				etn = int32(cl.Config.LibDefaults.PreferredPreauthTypes[0]) // Resort to config
 			}
-			// This is not in response to an error from the KDC. It is preemptive or renewal
-			et, err = crypto.GetEtype(etn) // Take the first as preference
+			et, err = crypto.GetEtype(etn)
 			if err != nil {
 				return krberror.Errorf(err, krberror.EncryptingError, "error getting etype for pre-auth encryption")
 			}
@@ -107,11 +107,12 @@ func setPAData(cl *Client, krberr messages.KRBError, ASReq *messages.ASReq) erro
 				return krberror.Errorf(err, krberror.EncryptingError, "error getting key from credentials")
 			}
 		}
-		// Generate the
+		// Generate the PA data
 		paTSb, err := types.GetPAEncTSEncAsnMarshalled()
 		if err != nil {
 			return krberror.Errorf(err, krberror.KRBMsgError, "error creating PAEncTSEnc for Pre-Authentication")
 		}
+		//TODO (theme: KVNO from keytab) the kvno should not be hard coded to 1 as this hampers troubleshooting.
 		paEncTS, err := crypto.GetEncryptedData(paTSb, key, keyusage.AS_REQ_PA_ENC_TIMESTAMP, 1)
 		if err != nil {
 			return krberror.Errorf(err, krberror.EncryptingError, "error encrypting pre-authentication timestamp")

--- a/client/client.go
+++ b/client/client.go
@@ -117,7 +117,7 @@ func (cl *Client) Key(etype etype.EType, krberr *messages.KRBError) (types.Encry
 	if cl.Credentials.HasKeytab() && etype != nil {
 		return cl.Credentials.Keytab().GetEncryptionKey(cl.Credentials.CName(), cl.Credentials.Domain(), 0, etype.GetETypeID())
 	} else if cl.Credentials.HasPassword() {
-		if krberr.ErrorCode == errorcode.KDC_ERR_PREAUTH_REQUIRED {
+		if krberr != nil && krberr.ErrorCode == errorcode.KDC_ERR_PREAUTH_REQUIRED {
 			var pas types.PADataSequence
 			err := pas.Unmarshal(krberr.EData)
 			if err != nil {

--- a/client/client.go
+++ b/client/client.go
@@ -180,10 +180,6 @@ func (cl *Client) Login() error {
 	if err != nil {
 		return krberror.Errorf(err, krberror.KRBMsgError, "error generating new AS_REQ")
 	}
-	err = setPAData(cl, messages.KRBError{}, &ASReq)
-	if err != nil {
-		return krberror.Errorf(err, krberror.KRBMsgError, "failed setting AS_REQ PAData")
-	}
 	ASRep, err := cl.ASExchange(cl.Credentials.Domain(), ASReq, 0)
 	if err != nil {
 		return err

--- a/keytab/keytab.go
+++ b/keytab/keytab.go
@@ -52,6 +52,7 @@ func New() *Keytab {
 
 // GetEncryptionKey returns the EncryptionKey from the Keytab for the newest entry with the required kvno, etype and matching principal.
 func (kt *Keytab) GetEncryptionKey(princName types.PrincipalName, realm string, kvno int, etype int32) (types.EncryptionKey, error) {
+	//TODO (theme: KVNO from keytab) this function should return the kvno too
 	var key types.EncryptionKey
 	var t time.Time
 	for _, k := range kt.Entries {
@@ -59,7 +60,6 @@ func (kt *Keytab) GetEncryptionKey(princName types.PrincipalName, realm string, 
 			k.Key.KeyType == etype &&
 			(k.KVNO == uint32(kvno) || kvno == 0) &&
 			k.Timestamp.After(t) {
-
 			p := true
 			for i, n := range k.Principal.Components {
 				if princName.NameString[i] != n {

